### DISCUSSION TITLE
The test time of the model on GPU is normal, but the test time on CPU is very long.

### DISCUSSION THREAD

**Discussion by xianyujie** [apache/incubator-mxnet#15108] [Q&A] [answered]
Hi, I trained the same model as the original model and only changed the datasets. When I test it On GPU, the test time of my model is similar to that of the original model, however, the time of my model tested On CPU is very larger than that of the original model. Why?

[a comment by an automated account is not shown]

**Comment by frankfliu** [comment]
@mxnet-label-bot add [Question]

**Comment by pengzhao-intel** [comment]
@xianyujie did you install the package of mxnet-cuXXmkl? 
The mkl build shows very great performance. Below two links give more details.

http://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu
https://mxnet.incubator.apache.org/versions/master/tutorials/mkldnn/MKLDNN_README.html

**Comment by xianyujie** [comment]
@pengzhao-intel the mkl build brings great performance to both models, but the issue still exits. 
By testing the run time of different network layers, I found that the problem occurred in most of the conv2_weight layers. The symbol.json of both models are same, and it's just that the value of arg_params is different, why does it make a huge difference in the running time of the model?

**Comment by pengzhao-intel** [comment]
Could you share the log or reproducible case for us?

**Comment by xianyujie** [comment]
Here, I tested the run time of different network layers on one image, I found that the problem occurred in most of the conv2_weight layers.
Tested on CPU, no mkl build, the first col shows my model, the second col shows the original model, the third col shows the network layer name.
[all_output_time_contrast.txt](https://github.com/apache/incubator-mxnet/files/3246126/all_output_time_contrast.txt)

**Comment by pengzhao-intel** [comment]
I don't find too much useful information from the log.
Could you give the step of reproducing the issue? Maybe by MXNet examples.

**Comment by xianyujie** [comment]
@pengzhao-intel  Can you find out what the problem is?             
**(The stage_unit1_relu1 is the front layer of the stage_unit1_conv1 layer)**

**===================The shape of the input image is (1,3,112,112)====================
warm up 20 times
same image as input,test the time from input layer to stage_unit1_relu1 layer**
(0.030597, 0.029112)
(0.029765, 0.027273)
(0.029243, 0.026952)
**same image as input,test the time from input layer to stage_unit1_conv1 layer**
(0.105118, 0.031303)
(0.098103, 0.031876)
(0.098488, 0.031119)
**same image as input, get the output(pre_output1,pre_output2) from stage1_unit1_relu1 layer 
of the two models.Then test the time of Conv layer, pre_output1 as the input of my model,
pre_output2 as the input of original model.**
(0.075896, 0.006333)
(0.07515, 0.005862)
(0.075065, 0.005814)
**The pre_output1 as the input of the two models, test the time of Conv layer**
(0.072311, 0.072548)
(0.073366, 0.075132)
(0.074931, 0.074962)
**The pre_output2 as the input of the two models, test the time of Conv layer**
(0.0055, 0.005653)
(0.005488, 0.005642)
(0.005499, 0.005644)
('loading bin', 0)
(2, 3, 12, 12)
**================The shape of the input image is (1,3,12,12)========================
warm up 20 times
same image as input,test the time from input layer to stage_unit1_relu1 layer**
(0.000733, 0.000735)
(0.000723, 0.000767)
(0.000761, 0.00073)
**same image as input,test the time from input layer to stage_unit1_conv1 layer**
(0.003494, 0.000923)
(0.003302, 0.000857)
(0.003363, 0.000906)
**same image as input, get the output(pre_output1,pre_output2) from stage1_unit1_relu1 layer 
of the two models.Then test the time of Conv layer, pre_output1 as the input of my model,
pre_output2 as the input of original model.**
(0.002808, 0.000264)
(0.002637, 0.000262)
(0.002527, 0.000257)
**The pre_output1 as the input of the two models, test the time of Conv layer**
(0.002551, 0.00255)
(0.002455, 0.002451)
(0.002425, 0.002443)
**The pre_output2 as the input of the two models, test the time of Conv layer**
(0.000261, 0.000258)
(0.000259, 0.000255)
(0.00026, 0.000255)

**Comment by pengzhao-intel** [comment]
It makes sense because your input image size is about 10X in both H and W direction from 12 to 112.
So the runtime is increased. Btw, the no mkl build will be very slow and it's not performant.

I suggest you change to MKLDNN build for the start point.

**Comment by xianyujie** [comment]
@pengzhao-intel I think you misunderstood the result. Take a look at the following results, different inputs have a great influence on the operation time of convolution layer, What could be the reason for this result?

**same image as input, get the output(pre_output1,pre_output2) from stage1_unit1_relu1 layer
of the two models.Then test the time of Conv layer, pre_output1 as the input of my model,
pre_output2 as the input of original model.**
(my_model: 0.075896, original model: 0.006333)
**The pre_output1 as the input of the two models, test the time of Conv layer**
(my_model: 0.072311, original model: 0.072548)
**The pre_output2 as the input of the two models, test the time of Conv layer**
(my_model: 0.0055, original model: 0.005653)

**Comment by pengzhao-intel** [comment]
Got it. It's interesting that the runtime is very different. 

Does the result of your models is correct?

**Comment by xianyujie** [comment]
@pengzhao-intel Yeah, I've tested it many times. Here is my test file link:
https://drive.google.com/open?id=1pzWZFY2cXsqYphit1HpMdGjntfSgIb9w

**Comment by xianyujie** [comment]
I've saved the pre_output1, the pre_output2, the conv1_weight, and the conv2_weight into the param file, and reload them to test.

**Comment by pengzhao-intel** [comment]
Thanks for the data and I can reproduce your issue now. 
Seems the convolution time is quick longer when you switched to the new dataset.

Debugging now and will back to you soon.

```
(base) [patric@mlt-skx132 test]$ python test.py
mkldnn_verbose,info,Intel(R) MKL-DNN v0.19.0 (Git Hash 41bee20d7eb4a67feeeeb8d597b3598994eb1959),Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with AVX512BW, AVX512VL, and AVX512DQ extensions
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nChw16c,num:1,1x64x112x112,0.628906
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,64x64x3x3,0.027832
mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:undef fdst:nChw16c,alg:convolution_direct,mb1_ic64oc64_ih112oh56kh3sh2dh0ph1_iw112ow56kw3sw2dw0pw1,**22.302**
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,64x64x3x3,0.027832
0.030622


(base) [patric@mlt-skx132 test]$ python test.py
mkldnn_verbose,info,Intel(R) MKL-DNN v0.19.0 (Git Hash 41bee20d7eb4a67feeeeb8d597b3598994eb1959),Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with AVX512BW, AVX512VL, and AVX512DQ extensions
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nChw16c,num:1,1x64x112x112,0.468994
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,64x64x3x3,0.0290527
mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:undef fdst:nChw16c,alg:convolution_direct,mb1_ic64oc64_ih112oh56kh3sh2dh0ph1_iw112ow56kw3sw2dw0pw1,**0.687012**
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,64x64x3x3,0.026123
0.009102


```

**Comment by pengzhao-intel** [comment]
After the analysis, I think the problem is the input data is underflow so the computation is very slow.
I think the problem will disappear and when you switch to a real data 

> [[[[-2.54311177e-28 -1.54346012e-29 -1.86278236e-27 ...  8.16160440e-02
>      3.19217145e-02  2.33461082e-01]
>    [-5.60695946e-27 -6.00699591e-27 -7.86982584e-27 ...  1.92388237e-01
>      9.94710028e-02  2.34758019e-01]
>    [-2.91797916e-27 -8.43729205e-28 -3.13466534e-27 ... -1.47056355e-27
>     -1.57520761e-27  1.33262098e-01]
>    ...
>    [ 1.02718532e-01 -4.38827914e-27 -3.15236521e-27 ... -5.44395992e-27
>     -7.84003017e-27 -8.97732492e-27]
>    [ 4.15135205e-01  4.14867938e-01  4.66439784e-01 ... -7.93092096e-27
>     -1.08695110e-26 -1.25687667e-26]
>    [-4.19972906e-27 -4.36781459e-27 -3.73858247e-27 ... -2.86622383e-27
>     -2.13053689e-27 -6.52746731e-27]]
> 
>   [[-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]
>    [-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]
>    [-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]
>    ...
>    [-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]
>    [-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]
>    [-6.91081895e-38 -6.91081895e-38 -6.91081895e-38 ... -6.91081895e-38
>     -6.91081895e-38 -6.91081895e-38]]

**Comment by xianyujie** [comment]
many thx, but the input data is obtained from the real data through many layers, and the results of all the real pictures are the same as above, so maybe the minimal calculation result is directly equal to 0?

**Comment by pengzhao-intel** [comment]
> many thx, but the input data is obtained from the real data through many layers, and the results of all the real pictures are the same as above, so maybe the minimal calculation result is directly equal to 0?

Make sense. I will contact the MKL-DNN team and see if any improvement can be done.

In a short time, please try to set the very small value to 0 :(

**Comment by xianyujie** [comment, marked as answer]
After I set the minimum value to 0, the test results of the two models are the same.

**same image as input, get the output(pre_output1,pre_output2) from stage1_unit1_relu1 layer 
of the two models.Then test the time of Conv layer, pre_output1 as the input of my model,
pre_output2 as the input of original model.**
(0.004521, 0.004521)
**The pre_output1 as the input of the two models, test the time of Conv layer**
(0.004198, 0.004238)
**The pre_output2 as the input of the two models, test the time of Conv layer**
(0.004196, 0.004184)
